client/memory.cpp

```diff
@@ -199,6 +228,24 @@ clEnqueueReadBufferRect(cl_command_queue command_queue, cl_mem buffer, cl_bool b
 	}
 }
 
+SO_EXPORT CL_API_ENTRY cl_int CL_API_CALL
+clEnqueueUnmapMemObject(cl_command_queue command_queue, cl_mem memobj, void* mapped_ptr,
+                        cl_uint num_events_in_wait_list, const cl_event* event_wait_list,
+                        cl_event* event) CL_API_SUFFIX__VERSION_1_0
+{
+	if (MappedObjects.find(memobj) == MappedObjects.end()) return CL_INVALID_VALUE;
+	if (MappedObjects[memobj].ptr != mapped_ptr) return CL_INVALID_VALUE;
+
+	cl_int ret;
+	ret = clEnqueueWriteBuffer(command_queue, memobj, true, MappedObjects[memobj].offset,
```

Comment:
Updating the server-side copy should only be done when the mapping has the WRITE permission set.